torch/distributed/_tensor/experimental/dmap.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+from typing import Callable, List, Optional, Sequence, Tuple, Union
+
+import torch
+from torch.distributed._tensor import DeviceMesh, DTensor
+from torch.distributed._tensor.placement_types import Placement
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+
+PlacementType = Sequence[Placement]
+InputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+OutputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+
+
+def dmap(
+    func: Callable,
+    out_placements: OutputPlacements,
+    device_mesh: Optional[DeviceMesh] = None,
+    in_placements: Optional[InputPlacements] = None,
+    resharding: bool = False,
+):
+    """
+    ``dmap`` is an experimental API that allows users to apply on :class:`DTensors`
+    a function that is written to be applied on :class:`~torch.Tensors`.
+
+    Args:
+        func (Callable): the function to be applied on each local shard of :class:`DTensor`s.
+        out_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]]):
+            the desired placements of the output :class:`DTensor`s.
+        device_mesh (:class:`DeviceMesh`, optional):
+            the device mesh that all the :class:`DTensor`s are placed on. If not specified,
+            this will be inferred from the input :class:`DTensor`s' device mesh. `dmap` requires
+            every :class:`DTensor`s to be placed on the same device mesh.
+        in_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]], optional):
```

Comment:
I think we should explicitly mention that the in_placements are out placements are flattened list if the function's actual inputs/outputs are complicated data structures like nested tuples or dict?